### PR TITLE
Create test_computetools.py with BorisPush tests

### PR DESCRIPTION
**Pull request by kphlips**
# Pull Request
### Description
Create test_computetools.py and create test for `BorisPush` `__init__` method and `push` method

This pull request addresses #97 

- [x] make sure you are pulling from a dev branch on your fork
- [x] check for code linting

### CODE REVIEW

**Comment by kphlips** [comment]
I'm not sure how to proceed given that the values from the calculation are not saved anywhere nor are they attributes.

File: tests/test_computetools.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+                  "type": "BorisPush",
+                  "Grid": {"N": 10, "min": 20, "max": 30},
+                  "PhysicsModules": {}}
+    owner = Simulation(input_data)
+    owner.prepare_simulation()
+    push_example = BorisPush(owner, input_data)
+    arr = np.ndarray(shape=(2, 2), dtype=float)
+    arr.fill(1)
+    push_example.push(arr, arr, 1, 1, arr, arr)
+    assert True
```
Comment (arichar6) [inline tests/test_computetools.py:28]: I'd do something like this:
```suggestion
    charge = 1.6022e-19      # proton charge in C
    mass = 1.6726e-27        # proton mass in kg
    E = np.zeros(3)
    B = np.zeros(3)
    x0 = np.array([[0, 0, 0.0]], dtype=np.float)
    v0 = np.array([[0, 0, 3.0e6]], dtype=np.float)  # velocity in m/s
    p0 = mass * v0              # nonrelativistic momentum (v<<c)
    dt = 1e-9                        # 1 ns timestep
    Nt = 10
    xfinal = v0 * Nt * dt
    for i in range(Nt):
        pusher.push(x0, p0, charge, mass, E, B)
    assert np.allclose(x0, xfinal)
```

Comment (arichar6) [reply tests/test_computetools.py:28]: The suggestion above checks that the position gets updated correctly if there are no fields. There should probably also be some checks for cases where the E and B fields are not zero. But we can do those later.

**Comment by kphlips** [comment]
Although implemented, this test returns false, stating `where False = <function allclose at 0x10ddd8f80>(array([[      0.        ,       0.        , 2999849.80105991]]), array([[0.  , 0.  , 0.03]]))
E        +    where <function allclose at 0x10ddd8f80> = np.allclose`

**Comment by arichar6** [comment]
I think that's because the time step for your clock is 0.1. Try changing the `dt` parameter in the dictionary that you pass to the Simulation constructor to 1e-9 instead of 0.1.

**Comment by kphlips** [comment]
Now the assertion seems to be caught up between [[0, 0, 0.299985]] and [[0, 0, 0.03]]

**Comment by arichar6** [comment]
you can change the tolerance for the `allclose` function, should help with passing the check

File: tests/test_computetools.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    clock = input_data["Clock"]
+    N = 10
+    xfinal = v_i * N * clock["dt"]
+    for i in range(N):
+        push_example.push(x_i, p_i, charge, mass, E, B)
+    assert np.allclose(x_i, xfinal, rtol=1e-4)
+    x_i = np.array([[0, 0, 0.0]], dtype=np.float)
+    p_i = 0 * v_i
+    E = np.array([[0, 0, 1e100]], dtype=np.float)
+    B = np.array([[0, 0, 1e100]], dtype=np.float)
```
Comment (arichar6) [inline tests/test_computetools.py:41]: FYI 1e100 is unrealistically large for E and B. Something on the order of 10 would be sufficient.

File: tests/test_computetools.py [hunk 3]
```diff
@@ -37,9 +37,9 @@ def test_push():
     assert np.allclose(x_i, xfinal, rtol=1e-4)
     x_i = np.array([[0, 0, 0.0]], dtype=np.float)
     p_i = 0 * v_i
-    E = np.array([[0, 0, 1e100]], dtype=np.float)
-    B = np.array([[0, 0, 1e100]], dtype=np.float)
+    E = np.array([[0, 0, 1e10]], dtype=np.float)
+    B = np.array([[0, 0, 1e10]], dtype=np.float)
```
Comment (arichar6) [inline tests/test_computetools.py:41]: I meant 10, not 1e10.
```suggestion
    E = np.array([[0, 0, 10]], dtype=np.float)
    B = np.array([[0, 0, 10]], dtype=np.float)
```

**Comment by kphlips** [comment]
@arichar6 Whoops, misunderstanding on my end!

File: tests/test_computetools.py [hunk 4]
```diff
@@ -1,8 +1,49 @@
...
+        push_example.push(x_i, p_i, charge, mass, E, B)
+    assert np.allclose(x_i, xfinal, rtol=1e-4)
+    x_i = np.array([[0, 0, 0.0]], dtype=np.float)
+    p_i = 0 * v_i
+    E = np.array([[0, 0, 10]], dtype=np.float)
+    B = np.array([[0, 0, 10]], dtype=np.float)
+    for i in range(N):
+        push_example.push(x_i, p_i, charge, mass, E, B)
+    check = np.array([[0, 0, 3.0]], dtype=np.float)
+    assert np.allclose(x_i, check, rtol=3e-2)
```
Comment (arichar6) [inline tests/test_computetools.py:44]:
```suggestion
def test_push_with_no_fields():
    """Tests the functionality of the push method, when E=B=0"""
    input_data = {"Clock":
                      {"start_time": 0, "end_time": 100, "dt": 1e-9},
                  "type": "BorisPush",
                  "Grid": {"N": 10, "min": 20, "max": 30},
                  "PhysicsModules": {}}
    owner = Simulation(input_data)
    owner.prepare_simulation()
    push_example = BorisPush(owner, input_data)
    charge = 1.6022e-19
    mass = 1.6726e-27
    E = np.zeros(3)
    B = np.zeros(3)
    x_i = np.array([[0, 0, 0.0]], dtype=np.float)
    v_i = np.array([[0, 0, 3.0e2]], dtype=np.float)
    p_i = mass * v_i
    clock = input_data["Clock"]
    N = 10
    x_final = v_i * N * clock["dt"]
    for i in range(N):
        push_example.push(x_i, p_i, charge, mass, E, B)
    assert np.allclose(x_i, x_final)


def test_push_with_crossed_fields():
    """Tests the functionality of the push method for E, B, not 0"""
    input_data = {
        "Clock": {"start_time": 0, "end_time": 100, "dt": 1e-9},
        "type": "BorisPush",
        "Grid": {"N": 10, "min": 20, "max": 30},
        "PhysicsModules": {}}
    owner = Simulation(input_data)
    owner.prepare_simulation()
    push_example = BorisPush(owner, input_data)
    charge = 1.6022e-19
    mass = 1.6726e-27
    N = 10
    x_i = np.array([[0, 0, 0.0]], dtype=np.float)
    p_i = np.array([[0, 0, 0.0]], dtype=np.float)
    E = np.array([[10, 0, 0]], dtype=np.float)
    B = np.array([[0, 0, 10]], dtype=np.float)
    for i in range(N):
        push_example.push(x_i, p_i, charge, mass, E, B)
    x_final = np.array([[2.20028479e-09, -1.04482737e-08, 0]])
    assert np.allclose(x_i, x_final)
    p_final = np.array([[0.47183691, -1.88168585,  0]]) * mass
    assert np.allclose(p_i / mass, p_final / mass)
```

Comment (arichar6) [reply tests/test_computetools.py:44]: Note the button that says "Commit suggestion". Here's a screenshot:
image
You can use that button to "accept" my proposed changes.

**Comment by arichar6** [comment]
Thanks for the contribution! Sorry this one was kinda troublesome.

**Pull request merged**